### ISSUE TITLE
Support for semver ranges?

### ISSUE THREAD

**Issue by jsumners** [nodejs/corepack#95]
I'd like to write a `package.json` like:

```json
{
  "packageJson": "pnpm@6"
}
```

Or maybe `pnpm@^6.32.2`, etc.

As it stands, adding a package manager to `package.json` requires a full specified version. This adds more maintenance overhead.

I think this is related to #93.

**Comment by shellscape** [comment]
Agreed. pnpm v7 was released just yesterday and I'd hoped to use corepack moving forward instead of manually installing it for all of our workflows. This is probably done so along the lines of "pin your dependencies," but doesn't meet all needs.

**Comment by styfle** [comment]
Would this need to perform a network request every time you do `pnpm run <script>` to check for the latest pnpm, since you can't rely the cached copy anymore?

**Comment by aduh95** [comment]
It looks like this was originally possible, and was removed in https://github.com/nodejs/corepack/pull/18.

**Comment by boneskull** [comment]
This seems kind of odd to restrict matching on exact version when one can provide any dist-tag instead (which can be moved around at will).  

Reverting the restriction would help my use case: use corepack to test or try something on a package manager where only the major version matters.  I have sinking feeling my use case is out of scope...

**Comment by aduh95** [comment]
@boneskull The restriction only applies to the `package.json` ad hoc field, you can still do e.g. `corepack pnpm@7.x install`, does that cover your use case?

**Comment by boneskull** [comment]
that wasn't working as of a couple hours ago. `corepack npm@7` unknown version or smth to that effect

**Comment by boneskull** [comment]
wrapping `corepack`  to support this, I've pulled down the result of `npm show --json <npm|yarn|pnpm> versions`, then use [semver](https://npm.im/semver) to try match a range.

**Comment by merceyz** [comment]
That seems like a bug `npm@7` isn't a range so it should expand to `npm@7.0.0`.

**Comment by boneskull** [comment]
@aduh95 I wonder if the env var was the problem.  Seems to work now, thanks.

**Comment by uiolee** [comment]
Totally agree. 

The exact version of the package manager prevents me from working on different projects. I don't think most projects need an exact package manager. corepack should introduce the semver strategy likes dependencies management.

**Comment by mrgrain** [comment]
I always want to use the latest version of my package manager. They exception usually is on a new major release, when I first have to run a few validations.

With the current restrictions to only exact versions, I have to make a change every time I want to update the version of the package manger. Worst case, this is includes a PR and I need to bother a co-worker with a review.

**Comment by boneskull** [comment]
@mrgain Couldn't you use a dist tag?  e.g., `npm@latest`

**Comment by mrgrain** [comment]
> @mrgrain Couldn't you use a dist tag?  e.g., `npm@latest`

I don't think dist tags are supported either? If they are, that wod certainly better than nothing.

However the problem with dist tags is

a) it's usually `latest` which will cause 7s to receive untested major version upgrades.

b) since a dist tag points to a specific version, we still impose an upgrade to the user.

**Comment by MattSilvaa** [comment]
Hi! I'm curious whether a decision has been reached to add the functionality back for supporting semver ranges?

**Comment by mcollina** [comment]
This would be great to implement this before calling it stable.

**Comment by arcanis** [comment]
I'm still not convinced this is a good idea:

- It prevents using hashes to ensure the version is the expected one
- It gives users a false sense of confidence that their project will always be installable ([bugfixes may be breaking changes](https://xkcd.com/1172/))

You can already use `corepack up` whenever you want to upgrade the package manager, I don't think an implicit upgrade should be encouraged as a good practice.

**Comment by mcollina** [comment]
What i would like the _default_ behavior to be given version of Node.js 22, is that corepack is enabled by default and `pnpm i` always installs the same major. I might have misunderstood this issue, but it would look like this is needed for that.

**Comment by mrgrain** [comment]
> I'm still not convinced this is a good idea:
> 
> * It prevents using hashes to ensure the version is the expected one
> * It gives users a false sense of confidence that their project will always be installable ([bugfixes may be breaking changes](https://xkcd.com/1172/))
> You can already use `corepack up` whenever you want to upgrade the package manager, I don't think an implicit upgrade should be encouraged as a good practice.

These are vary valid concerns and I understand that we should encourage explicit versions as the best practice. However I believe that many projects do not currently see this as a major concern, otherwise we would see a lot of very specific [engines](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#engines) declaration. 

The other use case is really around validation. If corepack already has a version of pnpm installed that satisfies the constraint `pnpm@6`, why should it install a slightly different version? This is just a waste of resources. Like you've said, users can run `corepack up` to upgrade. Of course explicit versions are still supported if you as owner prefer this. Ultimately this is about choice.

**Comment by xsjcTony** [comment]
+1

**Comment by Aeolun** [comment]
> if someone is that worried about that there should simply be a way of specifying an exact version

Yeah, this is common practice literally everywhere. If you care about getting exact versions, you specify the exact version. If you do not, you specify a range. I don't really understand how you can make a field that looks like it's a semver specifier, in a file full of semver specifiers, and then not actually make it semver. Like, wut?

**Comment by guilherssousa** [comment]
+1

**Comment by anthonyalayo** [comment]
What is the current best practice for this? Remove the entry in the package.json?

**Comment by aduh95** [comment]
> What is the current best practice for this? Remove the entry in the package.json?

You can run `corepack up` to bump the version in the same major, and `corepack use <your package manager>@<major version>.x` to use a different major version.

**Comment by jordanebelanger** [comment]
Not having semver makes this field so damn annoying when working as a team.

**Comment by karlhorky** [comment]
As of 12 Aug 2026 (`pnpm@11.21.0`), a fresh `pnpm init` project breaks immediately with Corepack because of this lack of support of version ranges:

```bash
$ ls -al $(which pnpm)
lrwxr-xr-x@ 1 k  admin  64 Jan 14  2026 /opt/homebrew/bin/pnpm -> ../Cellar/node@24/24.13.0/lib/node_modules/corepack/dist/pnpm.js

$ mkdir a

$ cd a

$ pnpm init
Wrote to /Users/k/p/a/package.json

{
  "name": "a",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devEngines": {
    "packageManager": {
      "name": "pnpm",
      "version": "^11.21.0",
      "onFail": "download"
    }
  },
  "type": "module"
}

$ pnpm -v
Invalid package manager specification in package.json (pnpm@^11.21.0); expected a semver version
```

Reported over here:

- https://github.com/pnpm/pnpm/issues/11732

@arcanis or other maintainers: what are your thoughts here? Would Corepack support version ranges, at least for pnpm?

Or is it no longer recommended to use Corepack with pnpm?

---

pnpm has now unfortunately dropped all mentions of Corepack from the docs because of this 🙁

- https://github.com/pnpm/pnpm/issues/11732#issuecomment-5267508169

---

## Workaround: Remove `devEngines` block from `pnpm init`

For people stuck with Corepack and pnpm (eg. before you have migrated off Corepack), there's a workaround:

```bash
$ pnpm init --init-package-manager false
Wrote to /Users/k/p/b/package.json

{
  "name": "b",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "type": "module"
}
```

Compare that with the full `pnpm init`:

```bash
$ pnpm init
Wrote to /Users/k/p/b/package.json

{
  "name": "b",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devEngines": {
    "packageManager": {
      "name": "pnpm",
      "version": "^11.21.0",
      "onFail": "download"
    }
  },
  "type": "module"
}
```
